README
README file for Module::Info::File

$Id: README,v 1.1 2003/09/28 09:25:04 jonasbn Exp $

In the beginning I was using Module::Info, but due to a lacking
functionality in this module I created Module::Info::File, which
inherits from Module::Info and replaces the B<new_from_file> method so
the lacking data can be accessed. Apart from that you can use all the
neat accessors from Module::Info.

In the bin folder in this distibution is a small script called
version.pl, which was the beginning of everything.

jonasbn, Copenhagen, Sun Sep 28 11:23:58 CEST 2003


CHANGES
Changes file for Module::Info::File

0.13 2014.08.25 maintenance release, update not necessary

- Additions to MANIFEST.SKIP

- License upgraded to Artistic 2.0 from Artistic 1.0


0.12 2014.08.23 maintenance release, update not necessary

- Specified Perl version to 5.6 as pointed out by Perl::Critic

- Ported from Module::Build to Dist::Zilla


0.11 2007.03.17, maintenance release, update not necessary

- Added t/perlcriticrc

- Updated t/pod.t, t/pod-coverage.t, t/kwalitee.t and t/critic
  to more contemporary versions

- Updated version number in POD


0.10 2007.02.13

- Bugfix release, a mistake I have made before, when letting
  Module::Build create the Makefile.PL is auto includes the bin
  directory, but with a full path referring to a path on the
  Build host.

  I have renamed bin/version.pl to script/version.pl and added
  it to Build.PL so not the Makefile.PL is generated correctly

  From Build.PL:

  scripts => {
      'script/version.pl'   => 'script/version.pl',
  },

  From Makefile.PL:

  'EXE_FILES' => [
        'script/version.pl'
   ],


0.09 2007.02.12

- Added a Perl::Critic test t/critic.t (courtesy of Jeffrey Ryan
  Thalhammer, Test::Perl::Critic)

- Refactored parts of the code, mostly regular expressions - all
  tests pass

- Ran code through perltidy aswell (level 1 perlcritic warning)

- Ran code through Perl::Critic all the way to level1 and
  updated POD and code accordingly after reading the referenced
  sections in PBP - they all made perfect sense

  Checked with Lars Thegler on the use of 2 vs. 3 argument
  version of open and since FreeBSD 4.x is declared EoL,
  Perl 5.6.x is the official version and compability for older
  perl is no longer required, if somebody would require this I
  would be willing to invest the time and effort

- Fixed a bug reported by  Eric D. Peterson, he experienced
  problems with traversing all his packages and getting the
  information using new_from_file()

  Apparently the DBD::Oracle has a closure in which the package
  is defined. See: Oracle::DBD, t/oracle_package_bug.t and
  drafts/OracleLikePackage.pm for an example

  I have changed the package matching regular expression,
  considering using PPI it might however be serious overkill.

- Changed from ExtUtils::MakeMaker to Module::Build

- Added a Kwalitee test t/kwalitee.t
  (courtesy of chromatic, Test::Kwalitee)

- Added a test more to t/File.t


0.08 2004.11.13

 - Improved the VERSION locating regular expression

 - Added use of Carp (also added in Makefile.PL)

 - Added t/00_load.t

 - Removed Multipackage files (no longer at part of the
   distribution)

 - new_from_file no longer supports list context, I have readded
   this to the TODO, since the implementation in 0.07 was not
   good, so multiple package definitions in files are no longer
   handled.

 - Fixed serious bug in the setting of the version number, this
   bug only occurred on Linux boxes running Perl 5.8.4.

 - Added some better exception handling to the version.pl script

 - Added some more debug information to the version.pl script

 - Turned this CHANGES file upside down

 - Added CPAN testers to the ACKNOWLEDGE section

 - Added version_bug.t to t/ (not in distribution)

0.07 Released 19th. of September 2004

- Added list context for constructor (new_from_file)

- bin/version.pl updated to accomodate this


0.06 2004.07.23

- Patched with patch from LTHEGLER, should now work under
  5.005_03, open was using the unecessary 3 arguments instead of just two.


0.05 2004.02.28

- Updated bin/version.pl with POD required to have the script in
  the CPAN scripts directory (http://www.cpan.org/scripts/). The
  script was uploaded alone to PAUSE together with the release
  of 0.04

- Added handling of directories so it is possible to get version
  numbers of all perl modules in a directory recursively with
  bin/version.pl

- Added INSTALL file

- Makefile.PL had a makeover


0.04 2004.01.10

- Updated bin/version.pl updated with error handling

- Update bin/version.pl with $VERSION number, I think this is a
  requirement for scripts in the scripts section on CPAN

- Added File::Spec to Makefile.PL

- Fixed a few bugs in bin/version.pl (I think I must have lost a
  version at some point)

- Added handling of .pl scripts

- Added a warning in the case where script is not a .pl or .pm

0.03 Released 20th. December 20 2003

- POD cleaned and reviewed by Rias.

- ACKNOWLEDGE section added

- Added information to t/Info.t so if the Module::Info tests
  fail, the test asks the tester to mail me the information


0.02 2003.11.03

- Added TODO file

- Added Changes file (this file)

- Added description to NAME section

- Added use of File::Basename to make test on Win32 work

- Fixed bug (rt: #4147) - broken tests on Win32


0.01 2003.09.30

Initial revision


script/version.pl
#!/usr/bin/perl -w

# $Id: version.pl 1707 2007-02-13 19:38:51Z jonasbn $

use strict;
use File::Find;
use Module::Info::File;
use Data::Dumper;
use diagnostics;
use vars qw($VERSION);

$VERSION = '0.05';

my $debug = 0;
my $modulepath;

if ($ARGV[0]) {
	$modulepath = $ARGV[0];
} else {
	usage();
}

my @ms;
if ($modulepath =~ m/::/) {
	print STDERR "Using new_from_module\n" if $debug;
	@ms = Module::Info::File->new_from_module($modulepath);
	foreach my $m (@ms) {
		long($m);
	}
} elsif (-f $modulepath) {
	print STDERR "Using new_from_file\n" if $debug;
	@ms = Module::Info::File->new_from_file($modulepath);
	foreach my $m (@ms) {
		long($m);
	}
} elsif (-d $modulepath) {
	find(\&simple, $modulepath);
} else {
	print STDERR "Falling back to new_from_module\n" if $debug;
	@ms = Module::Info::File->new_from_module($modulepath);
	foreach my $m (@ms) {
		long($m);
	}
}

unless(scalar(@ms)) {
	if (-d $ARGV[0]) {
		print "No modules/scripts located in: $modulepath\n";
	} else  {
		print "Unknown module/script: $modulepath\n";
	}
}

exit(0);

sub simple {
	if ($File::Find::name =~ m/\.pm$/) {

		print STDERR "Using new_from_file\n" if $debug;
		my @ms = Module::Info::File->new_from_file($_);
		foreach my $m (@ms) {
			print $m->name if $m->name;
			print ";";
			print $m->version if $m->version;
			print "\n";
		}
	}
}

sub long {
	my $m = shift;
	
	print STDERR Dumper $m if $debug;

	if ($m->isa("Module::Info") && $m->can("version")) {

		my $version = $m->version || 'N/A';

		print $m->name." located in ".$m->inc_dir." is version: ".$version."\n";
	}
}

sub usage {
	print "Usage: version.pl [<modulename>] [<modulepath>]\n";
	print "% perldoc version.pl for more information\n";
	exit(1);
}

__END__

=head1 NAME

version.pl - extracts module data from installed and uninstalled modules

=head1 SYNOPSIS

	% version.pl DBI
	
	% version.pl XML::Simple
	
	% version.pl ~jonasbn/Develop/Games/Bingo/lib/Games/Bingo.pm
	
	% version.pl version.pl
	
	% version.pl /System/Library/Perl/5.8.1

=head1 README

This script extracts module data from installed and uninstalled
modules and script, targetting especially the version information.

=head1 DESCRIPTION

The script takes either a module name (SEE SYNOPSIS), a path to a
Perl module, script file or a directory.

The script looks for a B<package> definition and a version variable.

The script is quite simple, the Module::Info::File and Mattia Barbon's
Module::Info holds all the interesting stuff.

In the beginning I was using Module::Info, but due to a lacking
functionality in this module I created Module::Info::File, which
inherits from Module::Info and replaces the B<new_from_file> method so
the lacking data can be accessed. Apart from that you can use all the
neat accessors from Module::Info. 

The script gives to kinds of input depending on how it was called:

For single modules, module og script files:

E<lt>modulenameE<gt> located in E<lt>directory locationE<gt> is
version: E<lt>versions numberE<gt>

Example:

Test::More located in /System/Library/Perl/5.8.1 is version: 0.47

Or when called with a directory as argument (one line per located module):

E<lt>module nameE<gt> E<lt>tabE<gt> E<lt>version numberE<gt>

Example:

DBI;1.38
Irssi;0.9
Mysql;1.2401
Bundle::DBI;11.03
Bundle::DBD::mysql;2.9002

=head1 SCRIPT CATEGORIES

Search

UNIX : System_administration

=head1 PREREQUISITES

=over 4

=item L<Module::Info::File 0.07>

=item L<Module::Info 0.20>

=item L<File::Basename>

=item L<File::Find>

=item L<File::Spec>

=item L<Data::Dumper>

=back

=head1 OSNAMES

any

=head1 SEE ALSO

=over 4

=item L<Module::Info>

=item L<Module::Info::File>

=back

=head1 AUTHOR

jonasbn E<lt>jonasbn@cpan.orgE<gt>

=head1 COPYRIGHT

version.pl is free software and is released under
the Artistic License. See
E<lt>http://www.perl.com/language/misc/Artistic.htmlE<gt> for details.

version.pl is (C) 2003-2004 Jonas B. Nielsen (jonasbn)
E<lt>jonasbn@cpan.orgE<gt>

=cut


lib/Module/Info/File.pm
package Module::Info::File;

use strict;
use warnings;
use base 'Module::Info';
use Carp;
use File::Spec;
use vars qw($VERSION);

$VERSION = '0.14';

sub new_from_file {
    my ( $proto, $filename ) = @_;

    open my $file, '<', $filename
        or croak "Unable to open file: $filename - $!";

    my ( @packages, $version, $name, $dir );
    while (<$file>) {
        if ( $_ =~ m/\bpackage\s?([A-Za-z0-9_:]+);/xm ) {
            if ($1) {
                $name = $1;
            } else {
                ($name) = $file =~ m/(\w+\.pm)$/xm;
            }
        }
        $dir = File::Spec->rel2abs($file);
        $dir =~ s[/(\w+\.p(m|l))$][]xm;

        if ( $_ =~ m/\bVERSION\b\s*=\s*'([0-9_.]+)'*;/xm ) {
            $version = $1;
        }

        if ( $name and $dir and $filename and $version ) {
            last;
        }

    }
    close $file;

    return __PACKAGE__->new_from_data(
        name    => $name,
        dir     => $dir,
        file    => $filename,
        version => $version ? $version : undef,
    );
}

sub new_from_data {
    my ( $class, %params ) = @_;

    my $self = bless {}, $class || ref $class;

    foreach ( keys %params ) {
        $self->{$_} = $params{$_} || undef;
    }

    if ( !$self->{'version'} ) {
        $self->{'version'} = $self->version();
    }

    return $self;
}

sub version {
    my ( $self, $version ) = @_;

    if ($version) {
        $self->{version} = $version;
        return 1;
    }

    if ( $self->{version} ) {
        return $self->{version};
    } else {
        return $self->SUPER::version();
    }
}

1;

__END__

=head1 NAME

Module::Info::File - retrieve module information from a file

=head1 VERSION

This POD describes version 0.13 of Module::Info::File

=head1 SYNOPSIS

	use Module::Info::File;

	my $module = Module::Info::File->new_from_file('path/to/Some/Module.pm');

	$mod->name();

	$mod->version();

	$mod->file();

	$mod->inc_dir();

=head1 DESCRIPTION

Module::Info (SEE REFERENCES), are lacking functionality of being able
to extract certain data when parsing a module directly from a file. I
have therefor created Module::Info::File, which inherits from
Module::Info and replaces the B<new_from_file> method so the lacking
data can be accessed (dir and name attributes). Apart from that you can
use all the neat accessors from Module::Info.

In the bin folder in this distribution is a small script called
version.pl, which was the beginning of everything.

=head1 SUBROUTINES/METHODS

=head2 new_from_file

Given a file, it will interpret this as the module you want information
about. You can also hand it a perl script.

After construction has been completed three attributes have been set in
the object:

=over 4

=item * name

=item * dir

=item * file

=back

So by using the inherited methods from B<Module::Info> you can access
the attributes.

There is an example in the bin/ folder called version.pl, this script
was the starting point for this module.

=over 4

=item * name, returns the package name of the file.

=item * version, returns the version number of the module/file in question
($VERSION).

=item * inc_dir, returns the dir attribute

=item * file, returns the file attribute

=back

=head2 new_from_data

A helper method to streamline the result set


In general please refer to the documentation on B<Module::Info> for more
details.

In list context the module returns and array of Module::Info::File objects, with
which you can use the above accessors. The information in the objects might not
be complete though (SEE: CAVEATS).

In the t/ directory of this distribution there is a test (Info.t), it
includes some tests. These tests will test your installation of
Module::Info (required by Module::Info::File), if the tests fail,
Module::Info::File will be obsolete and can be discontinued.

=head1 SEE ALSO

=over 4

=item * Module::Info, by Mattia Barbon

=item * bin/version.pl

=back

=head1 BUGS AND LIMITATIONS

The module can somewhat handle several package definitions in one file, but
the information is not complete yet, such as version information etc.

The method currently only support the following version number lines:

	$VERSION = '0.01';

	$VERSION = 0.01;

=head1 INCOMPATIBILITIES

Eric D. Paterson scanned his complete CPAN installation and got and came across
a few problems:

=over

=item L<DBD::Oracle>, the package was defined in a closure, this is handled from
Module::Info::File 0.09.

=item L<Archive::Zip::Tree>, which is installed, however deprecated and does NOT
contain a package defition.

=item L<CGI::Apache>, where the installed version was located in CGI/apache.pm
so the casing was not uniform with the package definition, I have asked Eric
for further information on this issue, regarding his version and installation.

=back

=head1 DIAGNOSTICS

=over

=item * Unable to open file: <filename> - <operating system error>

If the constructor B<new_from_file> is given a filename parameter, which does
not meet the following prerequisites:

=over

=item * it must be a file and it must exist

=item * the file must be readable by the user

=back

=back

=head1 CONFIGURATION AND ENVIRONMENT

There is a such no configuration necessary and Module::Info::File is expected
to work on the same environments as perl itself.

=head1 DEPENDENCIES

This module is a sub-class of L<Module::Info>, so the following direct
dependencies have to be met:

=over

=item L<Module::Info> 0.20

=item L<File::Spec>, to work on non-unix operating systems

=item Perl 5.6.0 and above

Compability for older perls is no longer required, if somebody would
require this I would be willing to invest the time and effort.

=back

=head1 TODO

=over

=item * Use PPI for parsing instead of own parsing mechanisms or it this serious
overkill?

=back

=head1 ACKNOWLEDGEMENTS

In no particular order

=over 4

=item *

Jeffrey Ryan Thalhammer, for L<Perl::Critic>

=item *

chromatic, for L<Test::Kwalitee>

=item *

Eric D. Peterson, for reporting bugs resulting in an improvement

=item *

Lars Thegler. tagg (LTHEGLER), for not letting me go easily, a patch and
suggesting the list context variation.

=item *

Thomas Eibner (THOMAS), for reviewing the POD

=item *

Mattia Barbon (MBARBON), for writing Module::Info

=item *

bigj at Perlmonks who mentioned Module::Info

=item *

All the CPAN testers, who help you to have your stuff tested on platforms not
necessarily available to a module author.

=back

=head1 AUTHOR

jonasbn E<lt>jonasbn@cpan.orgE<gt>

=head1 LICENSE AND COPYRIGHT

Module::Info::File and related modules are free software and is
released under the Artistic License 2.0.

Module::Info::File is (C) 2003-2014 Jonas B. Nielsen (jonasbn)
E<lt>jonasbn@cpan.orgE<gt>

=cut


Build.PL
# Build.PL for Module::Info::File

# $Id: Build.PL,v 1.15 2006/12/07 09:23:35 jbn Exp $

use strict;
use Module::Build;

my $build = Module::Build->new(
	dist_author       => 'Jonas B. Nielsen, <jonasbn@cpan.org>',
	dist_name         => 'Module-Info-File',
	dist_version_from => 'lib/Module/Info/File.pm',
    distribution_type => 'module',
	license           => 'perl',
	scripts => {
		'script/version.pl'   => 'script/version.pl',
	},
    recursive_test_files => 1,
	requires => {
		'Module::Info'   => '0.20',
		'File::Spec'     => 0,
		'Test::Harness'  => 0, #core
		'Test::More'     => 0, #core
		'Data::Dumper'   => 0, #core
		'UNIVERSAL'      => 0, #core
		'Carp'           => 0, #core
	},
    no_index => {
        directory => [ 'drafts', 't' ],
    },
	create_makefile_pl => 'traditional',
);
$build->create_build_script();

1;
